Add example for coupled physics (#377)
FreeFem/FreeFem-sources: examples/hpddm/coupled-dd-geneo-2d.md

---
name: coupledgeneo
category: Coupled Physics
layout: hpddm
---

## Coupled physics problem in 2d

In this example we will solve a coupled physics problem with two domains in 2d using domain decomposition, Krylov solvers and GenEO preconditioning. The setup consists of two domains that share an interface. On each domain a reaction-diffusion equation is defined,

$$
-\nabla u_1 + \eta u_1 = f_1\text{ on }\Omega_1\\

> -\nabla u_2 + \eta u_2=f_2\text{ on } \Omega_2

Reviewer: ```suggestion
-\nabla u_2 + \eta u_2 = f_2\text{ on }\Omega_2
```

$$
\alpha = \left[\left(\left(\frac{\pi}{L}\right)^2 + 1\right) \left(\left(\frac{\pi}{h}\right)^2 + 1\right)\right]^{\frac{1}{4}}
$$

### Weak form

We turn this into its weak form by using the usual test function approach, resulting in the definition of the local-to-domain problem,

$$
\mathscr{A}_i = \begin{bmatrix}L_{ii} & B_{ii}\\B_{ii}^T & C_{ii}\end{bmatrix}
$$

for $i=1,2$ as given by

$$
L_{ii} \leftarrow \int\int_{\Omega_i}(\nabla u_i \nabla v_i + \eta u_i v_i)\\
B_{ii} \leftarrow -\int_{\Gamma_i}p_i v_i \\
B_{ii}^T \leftarrow -\int_{\Gamma_i} u_i q_i \\
C_{ii} \leftarrow -\frac{1}{\alpha}\int_{\Gamma_i}u_i v_i.
$$

Similarly, the weak form for the cross-interface matrices between $\Omega_1$ and $\Omega_2$ are defined as

$$
B_{12} \leftarrow - \int_{\Gamma_1} p_2v_1\\
B_{21} \leftarrow - \int_{\Gamma_2} p_1v_2\\
C_{12} \leftarrow - \frac{1}{\alpha} \int_{\Gamma_1} p_2q_1\\
C_{21} \leftarrow - \frac{1}{\alpha} \int_{\Gamma_2} p_1q_2.
$$

### Preparing constants and building the mesh

First, we include the necessary headers and define some global constants and compute $\alpha$,

~~~freefem

> load "bem"

Author: indeed it is not (same for the `MUMPS_mpi` header. I forgot to remove them when trimming down my code into this example.

~~~

Next, we build a mesh (a star in this example) and split it into two halves along the line $x=0$. We also extract the interface meshes along the split:

~~~freefem
border star(t=0, 2*pi){
    x = 1.0*(1 + 0.3*sin(5*t))*cos(t);
    y = 1.0*(1 + 0.3*sin(5*t))*sin(t);
}
mesh Th = buildmesh(star(nn));
mesh Th1 = trunc(Th, x<=0, split=1, label=123);
mesh Th2 = trunc(Th, x>=0, split=1, label=123);
int[int] labs = [123];
meshL ThI1 = extract(Th1, label=labs);
meshL ThI2 = extract(Th2, label=labs);
~~~

### Domain decomposition

In order to use domain decomposition, we will take advantage of PETSc's capabilities through the `macro_ddm` script in FreeFem. However, the datatype `mesh` requires `macro_ddm` to be set up for a 2D volume mesh, whereas the `meshL` datatype requires a 3D line mesh setup. To accommodate this we will scope the relevant parts of the code doing the decomposition. To prepare for this, we need to prepare some variables that will get defined inside of those scoped regions but are also needed outside of them:

~~~freefem
Mat L11, L22;
Mat C11, C22;
Mat interpI12, interpI21;

fespace Ph1(Th1, P0);
fespace Ph2(Th2, P0);
Ph1 part1;
Ph2 part2;
~~~

> The `fespaces` defined will hold the decomposition which will allow us to reuse the decomposition of the surface mesh in the line mesh along the interface. Next, we decompose our surface meshes using a custom partitioning performed with `metis`, storing the partitioning in `part1` and `part2`:

Reviewer: ```suggestion
The `fespace`s defined will hold the decomposition which will allow us to reuse the decomposition of the surface mesh in the line mesh along the interface. Next, we decompose our surface meshes using a custom partitioning performed with `metis`, storing the partitioning in `part1` and `part2`:
```

real[int] rhs(Vh1.ndof + Vh2.ndof + VhI1.ndof + VhI2.ndof);
rhs = 0;
rhs(0:Vh1.ndof-1) = ProbL1(0,Vh1);
rhs(Vh1.ndof+VhI1.ndof:Vh1.ndof+VhI1.ndof+Vh2.ndof-1) = ProbL2(0,Vh2);
~~~

For simplicity, we define the same problem on both domains and use the same $f$ for both, however, this can be anything here.

### Constructing distributed matrices

We are now ready to set up and compose the various distributed block matrices that will make up the full system definition. We start with the matrices defining the problem local to $\Omega_1$ and $\Omega_2$,

~~~freefem
// L11 and L22
matrix mL11 = ProbL1(Vh1, Vh1, solver=GMRES);
matrix mL22 = ProbL2(Vh2, Vh2, solver=GMRES);
L11 = mL11;
L22 = mL22;
// B11 and B22
matrix mB11(Vh1.ndof, VhI1.ndof);
matrix mB22(Vh2.ndof, VhI2.ndof);
if(hasI1) mB11 = ProbB1(VhI1, Vh1, solver=GMRES);
if(hasI2) mB22 = ProbB2(VhI2, Vh2, solver=GMRES);
Mat B11(L11, C11);
Mat B22(L22, C22);
B11 = mB11;
B22 = mB22;
// C11 and C22
matrix mC11(VhI1.ndof,VhI1.ndof);
matrix mC22(VhI2.ndof,VhI2.ndof);
if(hasI1) mC11 = ProbC1(VhI1, VhI1, solver=GMRES);
if(hasI2) mC22 = ProbC2(VhI2, VhI2, solver=GMRES);
C11 = mC11;
C22 = mC22;
~~~

> Note that the value of the `solver` parameter is completely irrelevant as we will solve the system through PETSc later-on. However, the default solver in distributed FreeFem includes collective MPI operations, which does not allow us to only conditionally fill `B11`, `B22`, `C11`, and `C22`, as otherwise the execution would be stalled as soon as one of the subdomains does not include any part of the boundary.

Reviewer: ```suggestion
Note that the value of the `solver` parameter is completely irrelevant as we will solve the system through PETSc later-on. However, the default solver in distributed FreeFEM includes collective MPI operations, which does not allow us to only conditionally fill `B11`, `B22`, `C11`, and `C22`, as otherwise the execution would be stalled as soon as one of the subdomains does not include any part of the boundary.
```
